tools/ops/policystream.py

```diff
@@ -0,0 +1,610 @@
+    """
+    def __init__(self, repo_uri, repo, matcher=None):
+        self.repo_uri = repo_uri
+        self.repo = repo
+        self.policy_files = {}
+        self.matcher = matcher or policy_path_matcher
+
+    def initialize_tree(self, tree):
+        assert not self.policy_files
+
+        for tree_ent in tree:
+            fpath = tree_ent.name
+            if not self.matcher(fpath):
+                continue
+            self.policy_files[fpath] = PolicyCollection.from_data(
+                yaml.safe_load(self.repo.get(tree[fpath].id).data),
+                Config.empty(), fpath)
+
+    def delta_commits(self, baseline, target):
+        """Show policies changes between arbitrary commits.
+
+        The common use form is comparing the heads of two branches.
+        """
+
+        baseline_files = {}
+        for fent in baseline.tree:
+            if self.matcher(fent.name):
+                baseline_files[fent.name] = fent
+
+        target_files = {}
+        for fent in target.tree:
+            if self.matcher(fent.name):
+                target_files[fent.name] = fent
+
+        baseline_policies = PolicyCollection()
+        target_policies = PolicyCollection()
+
+        # Added
+        for f in set(target_files) - set(baseline_files):
+            target_policies += self._policy_file_rev(f, target)
+        # Removed
+        for f in set(baseline_files) - set(target_files):
+            baseline_policies += self._policy_file_rev(f, baseline)
+
+        # Modified
+        for f in set(baseline_files).intersection(target_files):
+            if baseline_files[f].hex == target_files[f].hex:
+                continue
+            target_policies += self._policy_file_rev(f, target)
+            baseline_policies += self._policy_file_rev(f, baseline)
+
+        return CollectionDelta(
+            baseline_policies, target_policies, target, self.repo_uri).delta()
+
+    def delta_stream(self, target='HEAD', limit=None,
+                sort=pygit2.GIT_SORT_TIME | pygit2.GIT_SORT_REVERSE):
+        """Return an iterator of policy changes along a commit lineage in a repo.
+        """
+        if target == 'HEAD':
+            target = self.repo.head.target
+
+        commits = []
+        for commit in self.repo.walk(target, sort):
+            log.debug(
+                "processing commit id:%s date:%s parents:%d msg:%s",
+                str(commit.id)[:6], commit_date(commit).isoformat(),
+                len(commit.parents), commit.message)
+            commits.append(commit)
+            if limit and len(commits) > limit:
+                break
+
+        if limit:
+            self.initialize_tree(commits[limit].tree)
+            commits.pop(-1)
+
+        for commit in commits:
+            for policy_change in self._process_stream_commit(commit):
+                yield policy_change
+
+    def _policy_file_rev(self, f, commit):
+        return PolicyCollection.from_data(
```

Comment:
there needs to be exception handling around this as policies may be checked into source control that are syntactically incorrect (which in some cases is expected, i.e. deprecation of filters/actions or mistakes)